aggregation/rustlib/src/user.rs

```diff
@@ -0,0 +1,290 @@
+
+use std::sync::Arc;
+use std::num::Wrapping;
+use std::collections::{BTreeMap, BTreeSet};
+
+use replace_with::*;
+use x25519_dalek;
+use sss_rs::wrapped_sharing::{Secret, share};
+use serde::{Serialize, Deserialize};
+use serde_json;
+
+use crate::sodium_bindings::*;
+use crate::helpers::*;
+use crate::types::*;
+
+fn round_0(data: &UserData) -> (OwnKeysData, (Signed<KAPublicKey>, Signed<KAPublicKey>)) {
+    let (comm_pk, comm_sk) = {
+        let secret = x25519_dalek::StaticSecret::new(rand_core::OsRng);
+        (x25519_dalek::PublicKey::from(&secret).to_bytes(), secret.to_bytes())
+    };
+    let (rand_pk, rand_sk) = {
+        let secret = x25519_dalek::StaticSecret::new(rand_core::OsRng);
+        (x25519_dalek::PublicKey::from(&secret).to_bytes(), secret.to_bytes())
+    };
+    let own_keys = OwnKeysData {
+        comm_pk,
+        comm_sk,
+        rand_pk,
+        rand_sk,
+    };
+
+    (own_keys, (Signed::wrap(comm_pk, &data.sign_sk), Signed::wrap(rand_pk, &data.sign_sk)))
+}
+
+fn round_1(
+    data: &UserData,
+    own_keys: OwnKeysData,
+    v: BTreeMap<usize, (Signed<KAPublicKey>, Signed<KAPublicKey>)>
+)
+    -> Result<((OwnKeysData, OthersKeysData, [u8; 32]), BTreeMap<usize, CryptoMsg>), ()>
+{
+    let n = v.len();
+    if n < data.threshold {
+        return Err(())
+    }
+
+    let i = v.iter()
+        .map(|(id, (x, y))| -> Option<_> { Some((data.others_sign_pks.get(id)?, (x, y))) })
+        .collect::<Option<Vec<_>>>()
+        .ok_or(())?;
+
+    if !(i.into_iter().all(|(pk, (x, y))| {
+        x.verify(pk).is_ok() && y.verify(pk).is_ok()
+    })) {
+        return Err(())
+    }
+
+    let comm_pks: BTreeMap<usize, KAPublicKey> = v.iter().map(|(id, (x, _))| (*id, *x.msg())).collect();
+    let rand_pks: BTreeMap<usize, KAPublicKey> = v.iter().map(|(id, (_, x))| (*id, *x.msg())).collect();
+
+    let seed = {
+        let mut seed = [0; 32];
+        getrandom::getrandom(&mut seed).map_err(|_| ())?;
```

Comment:
`// TODO: should investigate whether it makes more sense to block on /dev/random rather than letting /dev/urandom uninitialized CSPRNG provides unsuitable bytes`